docker/test_build_image_tagging.sh

```diff
@@ -0,0 +1,105 @@
+# $ ./build_image.sh -t "${TAG}" -g -py "${VERSION}" 
+# $ ./build_image.sh -g -py "${VERSION}" -t "${TAG}"
+# $ ./build_image.sh -g -t "${TAG}" -py "${VERSION}"
+# 
+# In order to assert the equivalence of all these variations, we take advantage
+# of how docker builds images: If two images are exactly the same (ie, they are
+# composed of the very same layers) they will have the same digest (ie, a hash 
+# value representing the content of the image), regardless of the tag assigned 
+# to the image. So, for example, if we run (with the same Dockerfile):
+# 
+# $ docker build -f Dockerfile -t Org/Repo:TagOne .
+# $ docker build -f Dockerfile -t Org/Repo:TagTwo .
+# $ docker images --no-trunc
+# 
+# we will see something like this:
+# 
+# REPOSITORY    TAG             IMAGE ID                                                               CREATED        SIZE                                   
+# Org/Repo     TagOne    sha256:e3824d794c0ccf10d2f61291f34e0d7e1e02e30b3d459465bc57d04dd3b65884    30 seconds ago   2.14GB
+# Org/Repo     TagTwo    sha256:e3824d794c0ccf10d2f61291f34e0d7e1e02e30b3d459465bc57d04dd3b65884    30 seconds ago   2.14GB
+# 
+# Notice that IMAGEID and CREATED are the same, since the first time it is 
+# actually created while the second time it just uses the cached layers. 
+# So the tag is "just a label" attached to the underlying image.
+#
+# Putting all together for our test:
+# We run `build_image.sh` (on the same machine to allow docker cache) with each
+# args order variation, tagging each variation with a different name (ensured 
+# by the random part of the string).
+# We expect:
+#   - All the tags to exist (tagging works): len(images_to_test) == len(tags_to_test)
+#   - All tagged images to be actually one and the same under the hood: len(set(digests)) == 1
+
+
+PY_VERSION=$1
+TAG_1="org/repo:image-${PY_VERSION}-${RANDOM}-${RANDOM}-${RANDOM}-${RANDOM}"
+TAG_2="org/repo:image-${PY_VERSION}-${RANDOM}-${RANDOM}-${RANDOM}-${RANDOM}"
+TAG_3="org/repo:image-${PY_VERSION}-${RANDOM}-${RANDOM}-${RANDOM}-${RANDOM}"
+TAG_4="org/repo:image-${PY_VERSION}-${RANDOM}-${RANDOM}-${RANDOM}-${RANDOM}"
+TAG_5="org/repo:image-${PY_VERSION}-${RANDOM}-${RANDOM}-${RANDOM}-${RANDOM}"
+TAG_6="org/repo:image-${PY_VERSION}-${RANDOM}-${RANDOM}-${RANDOM}-${RANDOM}"
+
+# Do builds alternating the flags order (-g, -t, -py)
+# (which should build only one underlying image)
+./build_image.sh -py "${PY_VERSION}" -t "${TAG_1}" -g
+./build_image.sh -py "${PY_VERSION}" -g -t "${TAG_2}"
+
+./build_image.sh -g -py "${PY_VERSION}" -t "${TAG_3}"
+./build_image.sh -g -t "${TAG_4}" -py "${PY_VERSION}"
+
+./build_image.sh -t "${TAG_5}" -py "${PY_VERSION}" -g 
+./build_image.sh -t "${TAG_6}" -g -py "${PY_VERSION}"
+
+# Collect all the images with their tags and ids
+IMGS_FILE="test_images.json"
+docker images --no-trunc --format "{{json .}}" | jq '{"repo": .Repository, "tag": .Tag, "digest": .ID}' | jq -s > "${IMGS_FILE}"
+
+python <<EOF
```

Comment:
Thanks @namannandan .  Hi @fabridamicelli this is perhaps another thing to consider for future PRs on adding more tests. It would be great if these can be implemented using pytest. You can take a look at the [this](https://github.com/pytorch/serve/tree/master/test/pytest) directory to see examples of how this is being done for the non-docker examples. We don't have an example with docker.

Reply:
Thank you for reviewing and for the feedback !

@namannandan
I can understand the point about the combination of python/bash code in one script. My logic behind having everything in one script is to kind of have an atomic component with (1 test -> 1 file) – and the mixture will anyways exist if we call things from python with `subprocess.run`. Still we could refactor it as you suggest and have it all in python.

@agunapal 
Regarding pytest: I take the point and I will check the other existing tests to follow the standards a bit closer, so we can step by step refactor things and iterate a bit on the design of the docker tests such that it is easy to extend with more tests and usage examples. I still have the feeling that it makes sense to keep docker tests under `docker/` for docker things to be all in one place..